test/OrchardCore.Tests/Modules/OrchardCore.Tenants/Controllers/AdminControllerTests.cs

```diff
@@ -0,0 +1,179 @@
+                .UpdateShellSettingsAsync(It.IsAny<ShellSettings>()))
+                .Callback<ShellSettings>(s => _shellSettings.Add(s));
+
+            _featureProfilesServiceMock = new Mock<IFeatureProfilesService>();
+            _featureProfilesServiceMock.Setup(fp => fp.GetFeatureProfilesAsync())
+                .Returns(Task.FromResult((IDictionary<string, FeatureProfile>)new Dictionary<string, FeatureProfile>
+                {
+                    { "Feature Profile", new FeatureProfile() }
+                }));
+        }
+
+        [Fact]
+        public async Task CreateTenantShouldChecksHostnameIfItAlreadyUsedInMultipleHostnames()
+        {
+            // Arrange
+            var shellSettingsManagerMock = new Mock<IShellSettingsManager>();
+            shellSettingsManagerMock.Setup(sm => sm.CreateDefaultSettings()).Returns(new ShellSettings());
+
+            var authServiceMock = new Mock<IAuthorizationService>(MockBehavior.Strict);
+
+            authServiceMock.Setup(x => x.AuthorizeAsync(It.IsAny<ClaimsPrincipal>(), It.IsAny<object>(), It.IsAny<IEnumerable<IAuthorizationRequirement>>()))
+                .Returns(Task.FromResult(AuthorizationResult.Success()));
+            authServiceMock.Setup(x => x.AuthorizeAsync(It.IsAny<ClaimsPrincipal>(), It.IsAny<object>(), It.IsAny<string>()))
+                .Returns(Task.FromResult(AuthorizationResult.Success()));
+
+            var stringLocalizerMock = new Mock<IStringLocalizer<AdminController>>();
+            stringLocalizerMock
+                .Setup(l => l[It.IsAny<string>()])
+                .Returns<string>(n => new LocalizedString(n, n));
+            stringLocalizerMock
+                .Setup(l => l[It.IsAny<string>(), It.IsAny<object[]>()])
+                .Returns<string, object[]>((n, a) => new LocalizedString(n, n));
+
+            var controller = new AdminController(
+                _shellHostMock.Object,
+                shellSettingsManagerMock.Object,
+                Enumerable.Empty<DatabaseProvider>(),
+                authServiceMock.Object,
+                _featureProfilesServiceMock.Object,
+                Enumerable.Empty<IRecipeHarvester>(),
+                Mock.Of<IDataProtectionProvider>(),
+                new Clock(),
+                Mock.Of<INotifier>(),
+                Mock.Of<ISiteService>(),
+                Mock.Of<IShapeFactory>(),
+                stringLocalizerMock.Object,
+                Mock.Of<IHtmlLocalizer<AdminController>>())
+            {
+                ControllerContext = new ControllerContext
+                {
+                    HttpContext = CreateHttpContext()
+                }
+            };
+
+            // Act & Assert
+            var result = await CreateTenantAsync(controller, "Tenant1", "example1.com, example2.com");
+
+            Assert.True(controller.ModelState.IsValid);
+
+            result = await CreateTenantAsync(controller, "Tenant2", "example1.com");
```

Comment:
Previously, this test was run with a lot of other variations of the input data as a data-driven test. I'd definitely keep that.

Reply:
> Otherwise, I like the refactoring. There were a couple of validation checks missing in the API previously.

Right

> I'd definitely keep that

I can bring them back, but the issue I can't controll the order of inline data to make sure everything as expected, that's why I didn't notice the NRE previously

Is there anything else or just bring those few tests back?

Reply:
I see, the order indeed matters. You could test multiple variations in the test then, with a deterministic order, without it being a data-driven one.

Reply:
That make sense